design/read-only-registry.md

```diff
@@ -0,0 +1,79 @@
+# Support for Read-Only Registries
+
+## Goals
+
+- Configure kotsadm to use application images from a private registry
+- Support using read-only registry during initial installation
+- Support using read-only registry on application updates
+- Support using read-only registry when settings are modified on the Registry Settings page in Admin Console
+- Support readon-only registries in airgap and on-line modes
+
+## Non Goals
+
+- Delivering images to the read-only registry.
+- Changing registry settings in airgap mode
+
+## Background
+
+Some production enviroments restrict image registry use to read-only mode.  This is an environment wide restriction.  Images are side-loaded into the registry from a trusted internal source.  Right now kotsadm will always push images to the private registry, making it impossible to use in such an environment.
+
+## Detailed Design
+
+### Install
+
+A new argument will be added to `kots install` command to flag the registry as a read-only registry.
+
+```
+kubectl kots install --kotsadm-registry my.private.registry/myapp \
+  --registry-username pulluser \
+  --registry-password pullpassword \
+  --registry-is-readonly \
+  app-slug
+```
+
+When the above command is used, the following will happen:
+
+1. All kotsadm pods will be rewritten to pull images from the private registry. This is existing functionality.
+1. Provided registry credentials will be saved in `.dockerconfigjson` secret in the app's namespace.
+1. During license upload in Admin Console UI, registry settings will be populated with these values.
+1. After the application is installed, Registry Settings page will be populated with these settings.
+
+Note: this changes current functionality where registry information specified in the command line arguments is not saved and is not applied to the application pods.
```

Comment:
Is anything else necessary to handle more edge cases, like what if the images haven't been pushed yet, or worse, some images have been pushed but not all of them? Usually when updating, we have reasonable assurance that the image pulls will work, so I wouldn't be surprised if our robustness will take a hit under this new flow. Do we need to do a preflight check to see if all the images actually exist before trying to use them?

Reply:
Going further, maybe we even need visual indication of whether all images exist or not. Or maybe we need a UI to list the images for someone to fetch out-of-band. Hopefully we don't need to go that far, but otherwise, how do people know what images to fetch?